rs/ledger_suite/icrc1/ledger/src/lib.rs

```diff
@@ -704,25 +707,45 @@ impl<Tokens: TokensType> Ledger<Tokens> {
     /// The canister code must call set_certified_data with the value this function returns after
     /// each successful modification of the ledger.
     pub fn root_hash(&self) -> [u8; 32] {
-        self.construct_hash_tree().digest().0
+        self.construct_hash_tree().digest()
     }
 
-    pub fn construct_hash_tree(&self) -> MixedHashTree {
+    pub fn construct_hash_tree(&self) -> HashTree {
         match self.blockchain().last_hash {
             Some(hash) => {
                 let last_block_index = self.blockchain().chain_length().checked_sub(1).unwrap();
-                MixedHashTree::Fork(Box::new((
-                    MixedHashTree::Labeled(
-                        Label::from("last_block_index"),
-                        Box::new(MixedHashTree::Leaf(last_block_index.to_be_bytes().to_vec())),
-                    ),
-                    MixedHashTree::Labeled(
-                        Label::from("tip_hash"),
-                        Box::new(MixedHashTree::Leaf(hash.as_slice().to_vec())),
-                    ),
-                )))
+                #[allow(unused_mut)]
+                #[allow(unused_assignments)]
+                let mut last_block_hash_label = Label::from("tip_hash");
+                let last_block_index_label = Label::from("last_block_index");
+
+                #[allow(unused_mut)]
+                #[allow(unused_assignments)]
+                let mut last_block_index_encoded = last_block_index.to_be_bytes().to_vec();
+                #[cfg(feature = "icrc3-compatible-data-certificate")]
+                {
+                    last_block_hash_label = Label::from("last_block_hash");
+                    let mut buf = std::io::Cursor::new(Vec::new());
+                    leb128::write::unsigned(&mut buf, last_block_index)
+                        .expect("Failed to write LEB128");
+                    last_block_index_encoded = buf.into_inner();
+                }
+
+                let mut tree_nodes = vec![
+                    (last_block_hash_label, leaf(hash.as_slice().to_vec())),
+                    (last_block_index_label, leaf(last_block_index_encoded)),
+                ];
+
+                // The nodes in the tree need to be in the correct order.
+                // Depending on whether icrc3 is enabled or not, the order of the nodes is different.
+                tree_nodes.sort_by(|a, b| a.0.cmp(&b.0));
+                let n = tree_nodes
+                    .into_iter()
+                    .map(|(l, n)| label(l, n))
+                    .collect::<Vec<_>>();
```

Comment:
If we were to have separate code blocks for the feature flag and without, and we construct the tree inside those blocks, we could also avoid this sorting IIUC.